astropy/io/fits/column.py

```diff
@@ -1389,19 +1389,21 @@ def __getattr__(self, name):
 
         Implements for example self.units, self.formats, etc.
         """
-
         cname = name[:-1]
         if cname in KEYWORD_ATTRIBUTES and name[-1] == 's':
             attr = []
-            for col in self:
+            for col in self.columns:
                 val = getattr(col, cname)
-                if val is not None:
-                    attr.append(val)
-                else:
-                    attr.append('')
+                attr.append(val if val is not None else '')
             return attr
         raise AttributeError(name)
 
+    def _update_attrs(self, name=None):
+        """Get or update local lists of column attributes"""
+        attrs = [name] if name is not None else KEYWORD_ATTRIBUTES
+        for name in attrs:
+            setattr(self, name + 's', self.__getattr__(name + 's'))
```

Comment:
Thanks for the timings.  I don't think sub-second speedups even for the more extreme cases are that convincing though...

But I could see making a special focus on names and formats.  In fact I think most of the code you've pointed to that is slowed down by this `.names` issue is not in the `ColDefs` class itself but in the `FITS_rec` class which uses it heavily.  `FITS_rec` already "subscribes" to change notifications from the `ColDefs`, so that would require no changes in `ColDefs`.

In fact, there is already a `FITS_rec.names` property, and really everywhere in `FITS_rec` that needs a list of the column names should use that (and for that matter `FITS_rec.names` and `ColDefs.<attr>s` should be tuples, not lists).  It used to be, some time ago, for various ugly reasons that `FITS_rec.names` and `ColDefs.names` might not be the same thing.  But now they are identical (in fact if you look at the listener for column name changes on `FITS_rec` it just updates the `.names` of the dtype). 

So really `FITS_rec.names` could just be a wrapper around `FITS_rec.dtype.names` and I think it should work if that is always used.  A similar approach could be taken for formats.


(unrelated: I noticed the `change_name` and `change_units` methods (which really should be deprecated IMO) call `_update_attrs` twice).